bundles/org.openhab.binding.pilight/src/main/java/org/openhab/binding/pilight/internal/discovery/PilightDeviceDiscoveryService.java

```diff
@@ -0,0 +1,215 @@
+                                .withThingType(thingTypeUID).withProperties(properties).withBridge(bridgeUID)
+                                .withRepresentationProperty(deviceId)
+                                .withLabel("Pilight " + typeString + " Device '" + deviceId + "'").build();
+
+                        thingDiscovered(discoveryResult);
+                    }
+                });
+            });
+
+            pilightBridgeHandler.refreshConfigAndStatus();
+        }
+    }
+
+    @Override
+    protected synchronized void stopScan() {
+        super.stopScan();
+        if (bridgeUID != null) {
+            removeOlderResults(getTimestampOfLastScan(), bridgeUID);
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Start Pilight device background discovery");
+        if (backgroundDiscoveryJob == null || backgroundDiscoveryJob.isCancelled()) {
+            backgroundDiscoveryJob = scheduler.scheduleWithFixedDelay(this::startScan, 20,
+                    AUTODISCOVERY_BACKGROUND_SEARCH_INTERVAL_SEC, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stop Pilight device background discovery");
+        if (backgroundDiscoveryJob != null) {
+            backgroundDiscoveryJob.cancel(true);
+            backgroundDiscoveryJob = null;
+        }
+    }
+
+    @Override
+    public void setThingHandler(final ThingHandler handler) {
+        if (handler instanceof PilightBridgeHandler) {
+            pilightBridgeHandler = (PilightBridgeHandler) handler;
+            bridgeUID = pilightBridgeHandler.getThing().getUID();
+        }
+    }
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return pilightBridgeHandler;
+    }
+
+    @Override
+    public void activate() {
+        super.activate(null);
+        if (pilightBridgeHandler != null) {
+            pilightBridgeHandler.registerDiscoveryListener(this);
+        }
+    }
+
+    @Override
+    public void deactivate() {
+        if (bridgeUID != null) {
+            removeOlderResults(getTimestampOfLastScan(), bridgeUID);
+        }
+
+        if (pilightBridgeHandler != null) {
+            pilightBridgeHandler.unregisterDiscoveryListener();
+        }
+
+        super.deactivate();
+    }
+
+    /**
+     * Method used to get pilight device config into the discovery class.
+     *
+     * @param config config to get
+     */
+    public void setConfig(Config config) {
+        if (!configFuture.isDone()) {
+            configFuture.complete(config);
+        }
+    }
+
+    /**
+     * Method used to get pilight device status list into the discovery class.
+     *
+     * @param status list of status objects
+     */
+    public void setStatus(List<Status> status) {
+        if (!statusFuture.isDone()) {
+            statusFuture.complete(status);
+        }
```

Comment:
There is no harm in trying to complete something that is already complete.
```suggestion
        statusFuture.complete(status);
```